src/databases.py

```diff
@@ -0,0 +1,108 @@
+# Copyright 2023 Canonical Ltd.
+# See LICENSE file for licensing details.
+
+"""Provide the Databases class to handle database relations and state."""
+
+import logging
+import typing
+
+from charms.data_platform_libs.v0.data_interfaces import DatabaseRequires
+
+from constants import FLASK_DATABASE_NAME, FLASK_SUPPORTED_DATABASES
+from exceptions import InvalidDatabaseRelationDataError
+
+if typing.TYPE_CHECKING:
+    from charm import FlaskCharm
+
+logger = logging.getLogger(__name__)
+
+
+class Databases:  # pylint: disable=too-few-public-methods
+    """A class handling databases relations and state.
+
+    Attrs:
+        _charm: The main charm. Used for events callbacks
+        _databases: A dict of DatabaseRequires to store relations
+    """
+
+    def __init__(self, charm: "FlaskCharm"):
+        """Initialize a new instance of the Databases class.
+
+        Args:
+            charm: The main charm. Used for events callbacks
+        """
+        self._charm = charm
+        self._databases: typing.Dict[str, DatabaseRequires] = {
+            name: self._setup_database_requirer(name, FLASK_DATABASE_NAME)
+            for name in FLASK_SUPPORTED_DATABASES
+        }
+
+    def _setup_database_requirer(self, relation_name: str, database_name: str) -> DatabaseRequires:
+        """Set up a DatabaseRequires instance.
+
+        The DatabaseRequires instance is an interface between the charm and various data providers.
+        It handles those relations and emit events to help us abstract these integrations.
+
+        Args:
+            relation_name: Name of the data relation
+            database_name: Name of the database (can be overwritten by the provider)
+
+        Returns:
+            DatabaseRequires object produced by the data_platform_libs.v0.data_interfaces library
+        """
+        database_requirer = DatabaseRequires(
+            self._charm,
+            relation_name=relation_name,
+            database_name=database_name,
+        )
+        self._charm.framework.observe(
+            database_requirer.on.database_created, self._charm.on_config_changed
```

Comment:
We discussed this for the Wordpress case as well, there shouldn't be these kind of circular dependencies between the Databases module and the charm. If the database relation change event needs to make changes at the charm level, then the observe initiation should be there, otherwise the event handling logic should only be on the Databases class.

Also, the ISD014 spec describes that there should be a 1-1 relationship between event and event handlers, so the `database_created` event should not use the `config_changed` event handler

If you would like to see an example of how a similar problem was solved, take a look at the Wordpress refactor spec or talk to Yanks 🙂

Reply:
This is also looking a lot like the observer pattern, it would be good not to re-invent it. We have discussed the Wordpress case during the sprint as a good way forward

Reply:
There's no simple way (IMHO) around the fact that we need to pebble replan to take changes to the flask layer into account.
Citing the current wp refactor spec: "Furthermore, if the event has changes that affect pebble configuration, the callback to the main charm class should be used".

> Also, the ISD014 spec describes that there should be a 1-1 relationship between event and event handlers, so the database_created event should not use the config_changed event handler

This would be a direct call to config_changed through this 3 lines (+docstrings) event handler. I love boilerplate as much as you but this seems excessive to me.

> This is also looking a lot like the observer pattern, it would be good not to re-invent it

I'm aware that it looks like it. But, as we discussed during our call, there's no accepted nor finalized version of the observer pattern for now. I'll be happy to change things in another PR but this seems out of scope for me right now.

Reply:
Given Gregory is back today, let's get him to take a look at this PR as well

Reply:
The other note, the Wordpress spec was discussed with Gregory and myself at the Prague sprint and these kind of issues came up and the spec was changed to address the issues. So just because it isn't finalised and approved doesn't mean that we should come up with other ways of doing the same thing which contains issues from an earlier version of the spec

Reply:
I think we shouldn't keep a call back to the charm from the module as it creates circular dependency. My proposal is to do a step further without implementing the full `Observer pattern` as discussed with Niels and Yanks. Indeed the full pattern would require some refactoring on parts of the charm which aren't in the scope of this task. We'll have a further task which will ask to refactor using the observer pattern and a task to generalise with a parent class to get ready for the 12-factorization of this project.

My proposal is to remove the callback and manage the hook in this module by doing something like:
```python
def on_database_created(self, event: DatabaseCreatedEvent):
   container = self.charm.unit.get_container(FLASK_CONTAINER_NAME)
   plan = container.get_plan()
   plan.services["myservice"].environment.update(self.get_uris())
   container.replan()
 ```
 
 Which removes the callback to the charm, while not being the pattern as we want

Reply:
Reworked